scripts/dump-openapi.py

```diff
@@ -44,17 +67,95 @@ def prefix_absolute_path_references(text, base_url):
     """
     return text.replace("](/", "]({}/".format(base_url))
 
-def edit_links(node, base_url):
-    """Finds description nodes and makes any links in them absolute."""
+def replace_match(text, match, replacement):
+    """Replaces the regex match by the replacement in the text."""
+    return text[:match.start()] + replacement + text[match.end():]
+
+def replace_shortcode(text, shortcode):
+    """Replaces the shortcode by a Markdown fallback in the text.
+    
+    The supported shortcodes are:
+
+    * boxes/note, boxes/rationale, boxes/warning
+    * added-in, changed-in
+    """
+
+    if shortcode['name'].startswith("/"):
+        # This is the end of the shortcode, just remove it.
+        return replace_match(text, shortcode['match'], "")
+    
+    match shortcode['name']:
+        case "boxes/note":
+            text = replace_match(text, shortcode['match'], "**NOTE:** ")
+        case "boxes/rationale":
+            text = replace_match(text, shortcode['match'], "**RATIONALE:** ")
+        case "boxes/warning":
+            text = replace_match(text, shortcode['match'], "**WARNING:** ")
+        case "added-in":
+            version = shortcode['params']['v']
+            if not version:
+                raise ValueError("Missing parameter `v` for `added-in` shortcode")
+        
+            text = replace_match(text, shortcode['match'], f"**[Added in `v{version}`]** ")
+        case "changed-in":
+            version = shortcode['params']['v']
+            if not version:
+                raise ValueError("Missing parameter `v` for `changed-in` shortcode")
+        
+            text = replace_match(text, shortcode['match'], f"**[Changed in `v{version}`]** ")
+        case _:
+            raise ValueError("Unknown shortcode", shortcode['name'])
+
+    return text
+
+
+def find_and_replace_shortcodes(text):
+    """Finds Hugo shortcodes and replaces them by a Markdown fallback.
+    
+    The supported shortcodes are:
+
+    * boxes/note, boxes/rationale, boxes/warning
+    * added-in, changed-in
+    """
+    # We use a `while` loop with `search` instead of a `for` loop with
+    # `finditer`, because as soon as we start replacing text, the
+    # indices of the match are invalid.
+    while match := shortcode_regex.search(text):
+        # Parse the parameters of the shortcode
+        params = {}
+        if match['params']:
+            for param in shortcode_params_regex.finditer(match['params']):
+                if param['key']:
+                    params[param['key']] = param['value']
+        
+        shortcode = {
+            'name': match['name'],
+            'params': params,
+            'match': match,
+        }
```

Comment:
Not a fan of this "`dict` as struct" usage. It makes Python into 1997-style Javascript and is generally hard to follow.

We've only got three properties here: just pass them separately into `replace_shortcode`? In fact, `name` is arguably redundant. 

Alternatively, we already have a dependency on [`attrs`](https://www.attrs.org/en/stable/) for these scripts, so define a class for them:

```
import attrs

@attrs.define
class ShortCodeInstance:
    name: str
    params: Dict[str, str]
    match: Match
```

Reply:
I removed it and passed just the match instead to `replace_shortcode`.